### ISSUE TITLE
vmalert running recording exhausted vmselect cpu and killed vmstorage 

### ISSUE THREAD

**Issue by renchili** [VictoriaMetrics/VictoriaMetrics#1702]
**Is your feature request related to a problem? Please describe.**
We have quite a lot of histograms or high cardinalities raw metrics. As parts of them came from open sources or unmodifiable third parties.  And in certain dimensions, the high cardinality label is the key label to distinguish the monitored system. 

So to let the query be faster than directly from vmstorage by users' query. We made some recording rules to reduce the cardinality of certain metrics and pre-calculated the PromSQL. But as we expanse in business, we add more and more targets to be scrapped. Now some times vmalert's query will be made vmselect CPU 100% busy.

And for more to have HA so we put up at least 2 vmalert server nodes. Then we found all the recording queries to be doubled to vmselect. And as we inserted 2 times of all data,  every time to query the vmselect metrics about recorded, vmselect needs to duplicate the job? 

**Describe the solution you'd like**
- allow vmalert like vmagent use cluster scope to shard between each other.  as vmagent use this: -promscrape.cluster.membersCount=1 -promscrape.cluster.replicationFactor=1 -promscrape.cluster.memberNum=0

**What more** 
as the realization about sharding system running a job will cost more than select direct the result. Is there any best practice to reduce the slow query?

**Comment by hagen1778** [comment]
> Is there any best practice to reduce the slow query?

There are multiple things to consider here:
1. Time series cardinality;
2. Amount of data to scan;
3. Transformations over scanned data.

The cardinality is basically meaning how many unique series VM needs to locate before giving a command to read them from disk. If rule expression queries a big number of time series (to check the number of series to be read by a rule just run query `count(metric_from_expression{with_filters})`), VM needs to check their locations in the index, which size depends on the total cardinality. If vmstorage doesn't have enough RAM to fit the index, it will take extra time to process it by chunks loading from disk (disk IO is always slow).

Amount of data to scan is the amount of bytes to read, decompress, unmarshall and process. There will be a noticeable performance difference between executing `rate(my_metric[5m])` and `rate(my_metric[1h])`: the former reads only 5min of recent data, when the latter needs 1h of data to read. And cardinality of `my_metric` is basically a multiplicator here.
Also, please note that reading recent data is usually much cheaper than reading old data because of page cache, which usually contains recently written/read data. Please also note, that page cache is used by most of the operations which touch the disk, so loading index into memory affects this too.

Transformations over the raw data read from disk. When vmstorage gets the data from the disk, it then sends it in response to vmselect. The latter is then responsible for processing, deduplication, executing functions, merging, etc. The CPU usage for vmselect depends on amounts of data it needs to process. So, for example, if it needs to process x2 time more data (due to deduplication) you can expect more CPU usage. Hence, deduplication enabled on vmstorage should reduce the CPU usage for vmselect.

Long story short, to reduce pressure on vm
